test/sqllogictest/regtype.slt

```diff
@@ -0,0 +1,223 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+reset-server
+
+# regtype
+
+query T
+SELECT 1::regtype
+----
+1
+
+query T
+SELECT 1::int4::regtype
+----
+1
+
+query T
+SELECT 1::oid::regtype
+----
+1
+
+query T
+SELECT 1::oid::regtype::oid
+----
+1
+
+query T
+SELECT '1'::regtype
+----
+1
+
+query T
+SELECT '1'::pg_catalog.regtype
+----
+1
+
+query T
+SELECT '1'::regtype::text
+----
+1
+
+query T
+SELECT 'date'::regtype::text
+----
+date
+
+query T
+SELECT 1082::regtype::text
+----
+date
+
+query T
+SELECT 'date'::regtype
+----
+1082
+
+query T
+SELECT 'date'::regtype::oid
+----
+1082
+
+query error type "dne" does not exist
+SELECT 'dne'::regtype
+
+query B
+SELECT 1082 = 'date'::regtype
+----
+true
+
+statement ok
+CREATE TABLE text_to_regtype (a text);
+
+statement ok
+INSERT INTO text_to_regtype VALUES (NULL), ('date');
+
+query I
+SELECT a::regtype FROM text_to_regtype ORDER BY a
+----
+1082
+NULL
+
+# Regression for 9194
+# This shouldn't be an error but seems to be impacted by
+# some evaluation order issue akin to #17189
+# TODO: this has an optimization bug.
+query error
+select 'date'::regtype::oid::text::regtype
```

Comment:
@MaterializeInc/compute This statement seems to have an optimization bug. If I
run it locally, I get one error, but if I run it through `sqllogictest`, I get a
different error. However, the error that results from when running in
sqllogictest seems to be a real bug of some variety.

For:

```
select 'date'::regtype::oid::text::regtype
```

If I run through `bin/environmentd`, I get:

```
Evaluation error: string is not a valid identifier: "1082"
```

This is #17189

If run through sqllogictest I get:

```
db error: ERROR: type "1082" does not exist: ERROR: type "1082" does not exist
```

This error comes from this PR
[here](https://github.com/MaterializeInc/materialize/pull/20407/files#diff-b945a38f3da415078dc2535b510ccea6e3d75cd0b441e34426418a3543cb702bR144).

However, if I break the query down to its components, everything works:

```
materialize=> select 'date'::regtype;
 regtype 
---------
 1082

select 1082::oid::text;
 text 
------
 1082

select 1082::text::regtype;
 regtype 
---------
 1082
```

What this looks like is that when the series of statements gets executed, [these
lines](https://github.com/MaterializeInc/materialize/pull/20407/files#diff-b945a38f3da415078dc2535b510ccea6e3d75cd0b441e34426418a3543cb702bR131-R132)
are getting skipped over. The `{2}` parameter is just a toggle [set to
`true`](https://github.com/MaterializeInc/materialize/pull/20407/files#diff-b945a38f3da415078dc2535b510ccea6e3d75cd0b441e34426418a3543cb702bR190)
for `regtype`. So we should be dealing with `$1` as the string `'1082'` cast to
`text`.

Looking at the plan that gets produced via `bin/environmentd`, I'm not seeing
anywhere that this gets misplanned as a `regclass`, so the check that the string
matches a number should always execute.

Is this just also #17189?